### ISSUE TITLE
Why is dist directory in source control?

### ISSUE THREAD

**Issue by Daniel15** [zenorocha/clipboard.js#79]
Generally, build artifacts should **not** be checked in to source control. This includes the `dist` directory in your repo. Out of curiosity, why did you decide to check this in? You can use the "releases" feature of Github to upload .zip files for people that don't want to use the npm package.

**Comment by jaydson** [comment]
https://github.com/zenorocha/clipboard.js/pull/48#issuecomment-144752014
>  There's no need to ignore .git and we can't remove dist because of Bower.

**Comment by zenorocha** [comment]
Yep, Bower is the reason. Thanks @jaydson.

**Comment by Daniel15** [comment]
Ahh. That sucks :( 

Bower is silly for requiring that. It would be nice to drop `dist` and Bower one day.

**Comment by FagnerMartinsBrack** [comment]
All of this is arguable and is totally up to the tradeoffs the author of a given project is willing to make. [The intention has never been for Bower to provide production-ready components](https://github.com/bower/bower/issues/368#issuecomment-15916138), but most projects, for conveniency, choose to do it.

Even if you want to provide production ready components for bower, you can leave the git repository without production files, but then the responsibility of publishing the "dist" folder would be entirely up to the automation of the release process, which takes time and resources some maintainers of some projects does not have.

---

**TL;DR:**
It is not a bower limitation, you can fix this providing minified files for bower and not for the git repo, but that demands time to adapt the release process and automate it.

---

There are [similar discussions](https://github.com/js-cookie/js-cookie/pull/83) in Open-Souce Land [out there](https://github.com/jquery/globalize/issues/210) along with [some alternatives](https://github.com/ReactiveX/RxJS/issues/90).

It would be very nice if someone could come up with an easy reusable pattern to solve this problem with Bower...

**Comment by Daniel15** [comment]
> you can fix this providing minified files for bower and not for the git repo, but that demands time to adapt the release process and automate it.

But you already need a build process like that for pushing to npm, I don't see how Bower would be much different. Most projects that push to npm already have a build target that handles that, a similar script could handle Bower too.

**Comment by FagnerMartinsBrack** [comment]
> But you already need a build process like that if pushing to npm, I don't see how Bower would be much different.

If you rely on tags it would definitely be the same. Bower does not have a registry, it is just an [Adapter](https://en.wikipedia.org/wiki/Adapter_pattern) to the original git repository, it doesn't even need `bower install` for each new version like npm. What you could do is commit the dist folder before releasing a tag to make it possible for people to download the minified file when using tags in bower version syntax (that's the kind of automation I was talking about).

> Most projects that push to npm already have a build target that handles that, a similar script could handle Bower too.

Not all projects publish production ready files to npm. Sometimes it makes sense to publish the minified files in a cdn like CDNJS and publish the source in npm, then you delegate the task of minification to the build system of the developer (if a project is downloading from npm, then it makes sense to assume that it has a build system than can minify it).

But as I said, it is all opinionated and project-specific because this is a general problem with no general solution. Sometimes ppl do as this project, which commit the production files in the "dist" folder for every change. Others commit only for the released tag. Others just delegate to the developer...

I would be pretty interested if anyone could come up (or recommend) a module that just works and does it all out of the box for both bower and npm.

**Comment by Daniel15** [comment]
> What you could do is commit the dist folder before releasing a tag to make it possible for people to download the minified file when using tags in bower version syntax

Off a branch I assume? If you do it against master you have the same issue (build artifacts checked in to the main branch). :stuck_out_tongue_closed_eyes: 

> Not all projects publish production ready files to npm

Sure, but many do.  I've seen quite a few libraries include both the source files (in `src` directory) and the compiled files (in `lib` or `dist` directory). The `lib` directory is **not** checked in to source control, it's only in the npm package. [Babel's default configuration is to ignore all files in `node_modules`](http://babeljs.io/docs/usage/require/#usage), so any npm package that uses ES6 features will immediately encounter issues out-of-the-box if the compiled version isn't in the npm package. An exception to that is when using modules in Node.js with ES6 features that are natively supported, but that's such a small use case compared to usage of JavaScript client-side (where you pretty much always need to compile to ES5).

The root of the problem is that the JavaScript ecosystem is extremely fragmented, which can make it difficult to work with. With something like PHP, there's a single package manager (Composer), a single main package repository (Packagist), and a single way to load any required third-party library (`require 'vendor/autoload.php'`). There's no need to try and work out which package manager you need to use or which module bundler is popular this week. Package managers in .NET (NuGet) and Python (PyPI) are pretty similar too. I'm hoping the JavaScript ecosystem reaches that level of maturity one day. Until then, we need to deal with a horribly fragmented ecosystem where every developer's configuration differs in very slight ways.

> I would be pretty interested if anyone could come up (or recommend) a module that just works and does it all out of the box for both bower and npm.

Not sure about Bower, but it's fairly simply to have a Gulp task that builds a library, packages it for npm, and pushes it. It's pretty much just:

1. Build everything into a temp directory (ie. copy over build artifacts as well as any other files you want to include in the package, like a readme)
2. Run `npm publish` in the temp directory

This is what I do for `HTMLtoJSX` for example: https://github.com/reactjs/react-magic/blob/master/gulpfile.js

**Comment by FagnerMartinsBrack** [comment]
> Off a branch I assume? If you do it against master you have the same issue (build artifacts checked in to the main branch).

I see no problem of doing it against `master`. As I understand the only problems for production files in the source repo are:

* Having unnecessary unstaged changes shown in a local branch when contributing to the project.
* The noise of production files when reviewing the diff of a PR.

If you assume people will be using tagged versions (or `#master`) in bower, my hypothesis would be to create a real commit with the files and another commit removing it after tagging. The "master" branch would never stay a long time in a state where the "dist" folder is visible, unless you checked out the exact commit in a fork (or checked out the latest in an unlucky timing).

It is not a silver bullet, but at least this is a valid pattern that would make it (kind of) work for both npm and bower.

**Comment by frankabbruzzese** [comment]
I am using globalize in asp.net core as a dependency of another bower package I implemented, and the fact that at moment there is no dist folder is a big problem, since it forces user to build globalize, and all its dependencies as a consequence of installing another package...
The best solution would be to give up bower and do everything with npm, but unluckly bower is the preferred way to install js packages to be deployed to the client while npm is the preferred package manager for development tools.
Now, since Asp.net core projects are pre-filled with the above logics, it would require a lot of extra work for the final user to use npm for client side packages. It would require also a lot of job to build manually bower packages...so this limit a lot the usage of the globalize library.
Is it so bad including dist folders? Practical needs are less important than "the theory" of avoiding built files in source control?
I think that this stuff is limiting the usage of the library, bringing the josb to having it working far aside what is acceptable. It is already fifficult to install and configure...also without this further difficulty, since the download and usage of cldr packages already add a difficulty level that is on the limit of acceptability for the average user.

**Comment by Daniel15** [comment]
>  the fact that at moment there is no dist folder is a big problem

You don't need to have dist checked in to your repo to handle Bower. You could either have a separate repo (eg. https://github.com/daniel15/babel-standalone-bower), or use npmcdn (which can load modules from NPM and package them so they work with Bower): https://npmcdn.com/

You can use NPM with ASP.NET Core projects with no issues. You can use a module bundler (such as Webpack or Browserify) and just make sure the resulting bundle is output to the `wwwroot` directory.
